File: claude_exp/routers/router_java/src/main/java/com/xv6/simulators/cryptohost/CryptoHostMain.java
package com.xv6.simulators.cryptohost;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpServer;
import com.xv6.shared.CommandServer;
import com.xv6.shared.IsoUtils;
import com.xv6.shared.LogLevels;
import com.xv6.shared.Stats;
import com.xv6.shared.StopEvent;

import java.io.File;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.logging.Level;
import java.util.logging.Logger;

/** Stateless HTTP service for cryptographic validation. Port of router_py's
 * simulators/crypto_host/main.py. */
public final class CryptoHostMain {

    private static final Logger logger = Logger.getLogger(CryptoHostMain.class.getName());
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final Map<String, Object> cfg;
    private final Map<String, Map<String, String>> pans;
    private final Stats stats;
    private final StopEvent stopEvent = new StopEvent();
    private final CommandServer cmd;

    public static Map<String, Object> loadConfig(String path) throws IOException {
        Map<String, Object> cfg = new LinkedHashMap<>(MAPPER.readValue(new File(path), Map.class));
        String baseDir = new File(path).getAbsoluteFile().getParent();
        cfg.put("pans_defined", resolve(baseDir, (String) cfg.get("pans_defined")));
        return cfg;
    }

    private static String resolve(String baseDir, String relative) {
        return Path.of(baseDir).resolve(relative).normalize().toString();
    }

    public CryptoHostMain(Map<String, Object> cfg) throws IOException {
        this.cfg = cfg;
        this.pans = MAPPER.readValue(
                new File((String) cfg.get("pans_defined")), new TypeReference<Map<String, Map<String, String>>>() {
                });

        Object yellow = cfg.get("yellow_threshold_seconds");
        this.stats = new Stats(yellow == null ? null : ((Number) yellow).intValue());
        Object cmdPort = cfg.get("command_port");
        this.cmd = new CommandServer(
                cmdPort == null ? 8082 : ((Number) cmdPort).intValue(),
                stats, stopEvent, "127.0.0.1", null);
    }

    // Stub: no PIN/ARQC/CVV2/AAV math - just reports whether the PAN is provisioned. Real
    // cryptographic validation lives in routers/crypto_host/ (the shared container); this stub
    // exists so router_java can still be built/tested standalone without it.
    private String validate(String pan, String f47Str) {
        Map<String, Object> data = IsoUtils.f47Decode(f47Str);
        data.put("response_code", pans.containsKey(pan) ? "00" : "14");
        return IsoUtils.f47Encode(data);
    }

    @SuppressWarnings("unchecked")
    private void registerFortanixRoute(HttpServer server, String pluginId, String bearerToken) {
        server.createContext("/sys/v1/plugins/" + pluginId, exchange -> {
            if (!"POST".equalsIgnoreCase(exchange.getRequestMethod())) {
                CommandServer.sendText(exchange, 405, "Method Not Allowed");
                return;
            }

            String authHeader = exchange.getRequestHeaders().getFirst("Authorization");
            if (authHeader == null || !authHeader.equals("Bearer " + bearerToken)) {
                CommandServer.sendJson(exchange, 401, Map.of("error", "unauthorized"));
                return;
            }

            try {
                stats.recordRecv();
                Map<String, Object> body = MAPPER.readValue(exchange.getRequestBody().readAllBytes(), Map.class);
                String operation = String.valueOf(body.getOrDefault("operation", ""));
                String pan = String.valueOf(body.getOrDefault("f2", ""));
                String f47 = String.valueOf(body.getOrDefault("f47", ""));
                // Not part of the Fortanix plugin contract - passed through purely so this
                // actor's logs can be joined with the router's logs on the same transaction
                // (empty when a caller doesn't send one, e.g. tests hitting this route directly).
                String routerStan = String.valueOf(body.getOrDefault("router_stan", ""));

                if (!operation.equals("validate_0100") && !operation.equals("validate_0110")) {
                    CommandServer.sendJson(exchange, 400, Map.of("error", "unknown operation: " + operation));
                    return;
                }

                logger.fine("validate pan=" + pan + " router_stan=" + routerStan);
                String result = validate(pan, f47);
                Map<String, Object> responseData = Map.of("f47", result);
                String responseJson = MAPPER.writeValueAsString(responseData);
                String base64Encoded = Base64.getEncoder().encodeToString(responseJson.getBytes(StandardCharsets.UTF_8));

                stats.recordSent();
                exchange.getResponseHeaders().set("Content-Type", "application/json");
                exchange.sendResponseHeaders(200, 0);
                exchange.getResponseBody().write(("\"" + base64Encoded + "\"").getBytes(StandardCharsets.UTF_8));
                exchange.close();
            } catch (Exception e) {
                logger.log(Level.WARNING, "Error processing request: " + e.getMessage());
                try {
                    CommandServer.sendJson(exchange, 500, Map.of("error", "internal error"));
                } catch (IOException ioe) {
                    logger.log(Level.WARNING, "Failed to send error response: " + ioe.getMessage());
                }
            }
        });
    }

    public void start() throws IOException {
        cmd.start();

        String pluginId = String.valueOf(cfg.getOrDefault("plugin_id", "default-plugin-id"));
        String bearerToken = String.valueOf(cfg.getOrDefault("bearer_token", "default-bearer-token"));

        int port = ((Number) cfg.get("port")).intValue();
        HttpServer validateServer = HttpServer.create(new InetSocketAddress("127.0.0.1", port), 0);
        registerFortanixRoute(validateServer, pluginId, bearerToken);
        logger.info("crypto host listening on port " + port);
        validateServer.setExecutor(Executors.newCachedThreadPool(runnable -> {
            Thread t = new Thread(runnable);
            t.setDaemon(true);
            return t;
        }));
        validateServer.start();
    }

    public void runForever() throws IOException, InterruptedException {
        start();
        stopEvent.await();
    }

    public static void main(String[] args) throws Exception {
        String configPath = null;
        for (int i = 0; i < args.length; i++) {
            if ("--config".equals(args[i]) && i + 1 < args.length) {
                configPath = args[i + 1];
            }
        }
        if (configPath == null) {
            throw new IllegalArgumentException("usage: CryptoHostMain --config <path>");
        }
        Map<String, Object> cfg = loadConfig(configPath);
        Logger.getLogger("").setLevel(LogLevels.parse("INFO"));
        try {
            new CryptoHostMain(cfg).runForever();
            // com.sun.net.httpserver.HttpServer's internal dispatcher thread is not a daemon
            // thread (confirmed: the process stayed alive after /stop with only that thread
            // left), so unlike router_py (where every thread genuinely is a daemon thread and the
            // interpreter exits on its own), this JVM needs an explicit exit once the stop event
            // has been honored.
            System.exit(0);
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: claude_exp/routers/router_cpp/src/simulators/crypto_host/crypto_host_main.cpp
#include <chrono>
#include <cstdlib>
#include <exception>
#include <memory>
#include <thread>

#include <nlohmann/json.hpp>

#include "router/router_config.h"
#include "shared/base64.h"
#include "shared/command_server.h"
#include "shared/iso_codec.h"
#include "shared/log.h"
#include "shared/pans_defined.h"
#include "shared/stats.h"
#include "shared/stop_event.h"

using json = nlohmann::json;
using namespace xv6::router;
using namespace xv6::shared;

namespace {

std::string parse_config_arg(int argc, char** argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (std::string(argv[i]) == "--config") {
            return argv[i + 1];
        }
    }
    throw std::runtime_error("missing --config <path>");
}

void send_json(httplib::Response& res, int status, const json& obj) {
    res.set_header("Content-Type", "application/json");
    res.set_content(obj.dump(), "application/json");
    res.status = status;
}

// Stub: no OpenSSL, no PIN/ARQC/CVV2/AAV math - just reports whether the PAN is provisioned.
// Real cryptographic validation lives in routers/crypto_host/ (the shared container); this stub
// exists so router_cpp can still be built/tested standalone without it.
json validate(const std::string& pan, const json& f47_in,
              const std::map<std::string, PanRecord>& pans) {
    json f47 = f47_in;
    f47["response_code"] = pans.find(pan) == pans.end() ? "14" : "00";
    return f47;
}

}  // namespace

int main(int argc, char** argv) {
    try {
        auto cfg_path = parse_config_arg(argc, argv);
        auto cfg = RouterConfig::from_file(cfg_path);
        auto pans = load_pans_defined(cfg.pans_defined_path);

        StopEvent stop_event;
        Stats stats(cfg.yellow_threshold_seconds);

        Logger::instance().set_level(*parse_log_level(cfg.log_level));
        CommandServer cmd(cfg.crypto.command_port, stats, stop_event, "0.0.0.0", {});
        cmd.start();

        // Separate httplib::Server for the plugin-execution route, bound to its own port --
        // mirrors the Java version's two independent HttpServer instances. SSLServer derives
        // from Server (same .Post/.listen/.stop surface), so ssl_active only decides which
        // concrete type gets constructed.
        std::unique_ptr<httplib::Server> plugin_server_ptr;
        if (cfg.crypto.ssl_active) {
            plugin_server_ptr = std::make_unique<httplib::SSLServer>(
                cfg.crypto.certfile.c_str(), cfg.crypto.keyfile.c_str(),
                cfg.crypto.cafile.empty() ? nullptr : cfg.crypto.cafile.c_str());
        } else {
            plugin_server_ptr = std::make_unique<httplib::Server>();
        }
        httplib::Server& plugin_server = *plugin_server_ptr;
        plugin_server.Post(
            "/sys/v1/plugins/([^/]+)",
            [&](const httplib::Request& req, httplib::Response& res) {
                std::string plugin_id = req.matches[1];
                if (plugin_id != cfg.crypto.plugin_id) {
                    res.status = 404;
                    return;
                }
                if (req.get_header_value("Authorization") != "Bearer " + cfg.crypto.bearer_token) {
                    send_json(res, 401, {{"error", "unauthorized"}});
                    return;
                }

                // router_stan isn't part of the Fortanix plugin contract - it's an extra field
                // the router sends purely so this actor's logs can be joined with the router's
                // logs on the same transaction (empty when a caller doesn't send one, e.g. tests
                // hitting this route directly).
                std::string router_stan;
                try {
                    auto body = json::parse(req.body);
                    std::string pan = body.at("f2").get<std::string>();
                    std::string f47_str = body.value("f47", std::string(""));
                    router_stan = body.value("router_stan", std::string(""));
                    json f47 = iso_codec::f47_decode(f47_str);
                    stats.record_recv();

                    json enriched = validate(pan, f47, pans);
                    LOG_DEBUG("validated pan=" + pan + " router_stan=" + router_stan);

                    json envelope = {{"f47", enriched.dump()}};
                    std::string envelope_str = envelope.dump();
                    std::string b64 = base64_encode(
                        std::vector<uint8_t>(envelope_str.begin(), envelope_str.end()));

                    res.set_header("Content-Type", "application/json");
                    res.set_content(json(b64).dump(), "application/json");
                    res.status = 200;
                    stats.record_sent();
                } catch (const std::exception& e) {
                    LOG_WARNING("validate failed router_stan=" + router_stan + ": " + e.what());
                    send_json(res, 400, {{"error", e.what()}});
                }
            });

        std::thread plugin_thread([&] { plugin_server.listen("0.0.0.0", cfg.crypto.port); });
        LOG_INFO("crypto_host listening on port " + std::to_string(cfg.crypto.port));

        while (!stop_event.is_set()) {
            stop_event.wait_for(std::chrono::seconds(1));
        }

        plugin_server.stop();
        if (plugin_thread.joinable()) plugin_thread.join();
        cmd.stop();
        return 0;
    } catch (const std::exception& e) {
        LOG_ERROR(std::string("fatal: ") + e.what());
        return 1;
    }
}
